### PR TITLE
Closes #6414: Enable assigning prefixes to various object types

### LINKED ISSUE #6414: Allow prefixes to be assigned to a site group, region, site, or location
### NetBox version

v2.11.3

### Feature type

New functionality

### Proposed functionality

Allow prefixes to be assigned to a site-group and not a specific site

### Use case

Lets say we have a site-group called "Singapore" which contains multiple sites e.g SG1, SG2, SG3
We have servers in each site which get assigned /31s IPs from a /24
So we'd like the ability to assign the /24 prefix to the site-group only, since it does not "belong" to only one site.
From there we can assign smaller prefixes into the relevant site as needed

### Database changes

unsure, don't think so

### External dependencies

none

### PR DESCRIPTION
**Pull request by jeremystretch**
### Fixes: #6414

- Add a `scope` GenericForeignKey on Prefix
- Add denormalized relations on Prefix to the scope object types (region, site group, site, and location)
- Populate scope & denormalized relations on migration
- Remove the original `site` ForeignKey field from Prefix
- Update PrefixFilterSet filters

### CODE REVIEW

File: netbox/ipam/forms/bulk_edit.py [hunk 1]
```diff
@@ -203,25 +203,6 @@ class RoleBulkEditForm(NetBoxModelBulkEditForm):
 
 
 class PrefixBulkEditForm(NetBoxModelBulkEditForm):
```
Comment (arthanson) [inline netbox/ipam/forms/bulk_edit.py:205]: should we add scope here?

File: netbox/ipam/forms/filtersets.py [hunk 2]
```diff
@@ -224,12 +224,13 @@ class PrefixFilterForm(TenancyFilterForm, NetBoxModelFilterSetForm):
     site_id = DynamicModelMultipleChoiceField(
         queryset=Site.objects.all(),
         required=False,
-        null_option='None',
-        query_params={
-            'region_id': '$region_id'
-        },
         label=_('Site')
     )
+    location_id = DynamicModelMultipleChoiceField(
```
Comment (arthanson) [inline netbox/ipam/forms/filtersets.py:229]: wondering if these could be a mixin ScopedFilterSet as we are going to be using this in several models and they will all have the same filter, form fields etc.. could have site_id, location_id, site_group all in the mixin.

File: netbox/ipam/forms/model_forms.py [hunk 3]
```diff
@@ -228,17 +234,48 @@ class PrefixForm(TenancyForm, NetBoxModelForm):
...
 
+    def __init__(self, *args, **kwargs):
+        instance = kwargs.get('instance')
+        initial = kwargs.get('initial', {})
+
+        if instance is not None and instance.scope:
+            initial['scope'] = instance.scope
+            kwargs['initial'] = initial
+
+        super().__init__(*args, **kwargs)
```
Comment (arthanson) [inline netbox/ipam/forms/model_forms.py:257]: again seems like a good potential place for a mixin to reduce boilerplate as we are going to be using this same pattern in about 5 places.

**Review by arthanson — CHANGES_REQUESTED** [review]

File: netbox/ipam/forms/bulk_edit.py (hunk 1 above)
Comment (jeremystretch) [reply netbox/ipam/forms/bulk_edit.py:205]: I think we run into a snag using HTMXSelector on a bulk edit form: The bulk edit forms must be POSTed to capture the selected object PKs, whereas HTMXSelector uses a GET to re-render the form. We'll need to extend HTMXSelector to allow POSTing instead.

File: netbox/ipam/forms/filtersets.py (hunk 2 above)
Comment (jeremystretch) [reply netbox/ipam/forms/filtersets.py:229]: It's not yet clear which models will support this same set of scope types. VLANGroup, for instance, supports these four but also racks, cluster groups, and clusters. We'll utilize mixins at some point for sure but it's probably still too early for that degree of optimization.

File: netbox/ipam/forms/bulk_edit.py (hunk 1 above)
Comment (jeremystretch) [reply netbox/ipam/forms/bulk_edit.py:205]: Going to work on #17789 first to get a pattern established in `develop`, then port it over here.

**Comment by jeremystretch** [comment]
Temporarily blocked by some prework for GFK bulk edit support in #17789

**Pull request merged**